arch: common: multilevel_irq: register directly from driver

ycsin opened the pull request:
#63172 reworked the LUT struct in the `multilevel_irq.c` (formerly `sw_isr_table.c`) to have an additional `struct device` pointer member so that Zephyr can know the controller responsible for an interrupt request by referring to the table.

![Screenshot 2024-04-18 at 10 27 57 AM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/357442dd-de97-4bb3-a438-488e7375ad6b)

The tables generation implementation in #63172 continues to be done directly in the `multilevel_irq.c`, however, it has no knowledge of the interrupt controllers and have to rely on lots of macro magic🧙‍♂️✨ to:

1. Loop through the devicetree, and search for interrupt controllers by looking for the `interrupt-controller` property
2. Find the interrupt controller which IRQ is equals to aggregator interrupt offset defined by the Kconfigs (i.e. `CONFIG_2ND_LVL_INTR_00_OFFSET`)

which is both flimsy and hard to understand.

This PR seeks to flip things around, that is, allow the interrupt controller(s) to register itself with the multilevel LUT by using the `IRQ_PARENT_ENTRY_DEFINE` macro. Additionally, `INTC_ISR_TBL_OFFSET` is made to get the SW ISR table level offset based on the order of the controller.

The LUT, now includes an additional `level` member, is placed directly into the iterable struct section, eliminating the need to have 2 separate tables and additional macro guards in the `multilevel_irq.c`.

This implementation reduces the dependencies on some of the Kconfigs in the `kconfig.multilevel`, the remaining Kconfigs are still required until `gen_isr_table.py` understands the devicetree.

![Screenshot 2024-04-18 at 10 28 11 AM](https://github.com/zephyrproject-rtos/zephyr/assets/1563974/9289903e-9b1b-46a9-aa48-4627b10ccc29)

[Migration doc preview](https://builds.zephyrproject.io/zephyr/pr/66877/docs/releases/migration-guide-3.7.html)

Fixes #72979

ycsin committed:
arch: sw_isr: revamp multi-level interrupt architecture

Previously the multi-level irq lookup table is generated by
looping through the devicetree nodes using macros & Kconfig,
which is hard to read and flimsy.

This PR shifts the heavy lifting to devicetree & DT macros such
that an interrupt controller driver, which has its info in the
devicetree, can register itself directly with the multi-level
interrupt architecture, which is more straightforward.

The previous auto-generated look up table with macros is now
moved in a file of its own. A new compatibility Kconfig:
`CONFIG_LEGACY_MULTI_LEVEL_TABLE_GENERATION` is added and
enabled by default to compile the legacy look up table for
interrupt controller drivers that aren't updated to support the
new architecture yet.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: simplify isr handling

Save the reference of the start index of the `_sw_isr_table`
to the config struct, so that the `local_irq` can be used  as
offset directly.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented:
@ycsin - please rebase

ycsin committed:
doc: migration-guide: 3.7: add note on multilevel interrupt arch changes

Add a note on the revamp of the multilevel interrupt
architecture, the changes to be made, and new macros to help
with the update.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on scripts/dts/gen_defines.py:
```diff
@@ -499,7 +499,6 @@ def map_arm_gic_irq_type(irq, irq_num):
     for macro, val in name_vals:
         out_dt_define(macro, val)
 
-
```
Might as well leave that blank line where it is

cfriedt commented on cmake/linker_script/common/common-rom.cmake:
```diff
@@ -202,6 +202,10 @@ if (CONFIG_LOG)
 endif()
 
+if (CONFIG_MULTI_LEVEL_INTERRUPTS)
+  zephyr_iterable_section(NAME intc_table KVMA RAM_REGION GROUP RODATA_REGION SUBALIGN 4)
```
I wonder if it would make sense to have 3 iterable sections - one for each level 🤔

ycsin replied:
One of the earlier version has that, with different tables for each level we can remove `level` from the `struct _irq_parent_entry` but we need multiple `STRUCT_SECTION_FOREACH_ALTERNATE`s for each of the tables in `multilevel_irq.c`

cfriedt replied:
> we need multiple `STRUCT_SECTION_FOREACH_ALTERNATE`s for each of the tables in `multilevel_irq.c`

Alternatively, you could just create a wrapper (static inline function or macro) that would accept `level` as a parameter and iterate only over that particular table.

cfriedt replied:
@ycsin - should here be additional sections defined here? I can't remember what the outcome was.

cfriedt reviewed (commented):
Overall LGTM - just a couple of questions / nits.

I like that there is significant code removal 😁

ycsin commented:
cc @LaurentiuM1234(intc_nxp_irqsteer.c)  @dleach02(intc_rv32m1_intmux.c)  @kv2019i(intc_cavs.c)

ycsin commented:
> I like that there is significant code removal 😁

They are now relocated into a separate file queueing for deprecation instead of outright removal 😬

ycsin commented on arch/common/multilevel_irq_legacy.c:
```diff
@@ -0,0 +1,72 @@
+				   CONFIG_2ND_LVL_INTR_0##i##_OFFSET,                              \
+				   IRQ_INDEX_TO_OFFSET(i, base))
+
+LISTIFY(CONFIG_NUM_2ND_LEVEL_AGGREGATORS, CAT_2ND_LVL_LIST, (;), CONFIG_2ND_LVL_ISR_TBL_OFFSET);
```
Updated to make use of the new `IRQ_PARENT_ENTRY_DEFINE` macro instead of:

https://github.com/zephyrproject-rtos/zephyr/blob/a06b7a4676962b39a8b8ccdcaf343c964a0ce86b/arch/common/multilevel_irq.c#L59-L61

so that the implementations in `arch/common/multilevel_irq.c` can be kept as-is

ycsin committed:
irq: multilevel: add APIs with `level` as argument

Add APIs that accept an additional `level` argument, and use
that to call the respective functions. In some cases this can
reduce code complexity, especially when the level isn't known
at build time.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
arch: common: multilevel_irq: simplification with new multilevel IRQ APIs

Use the multi-level interrupt APIs that accepts `level` as an
argument for the code where the level of the interrupt is not
known at build time.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Could you take a look at this one please? This is essentially a cleanup to the existing multilevel implementation that which I hope that it can make it to the v3.6 release so that we don't carry the old implementation for too long. I'm not sure if this needs to go thru an arch meeting - @andyross @peter-mitsis @dcpleung @fkokosinski @carlocaione

bjarki-andreasen reviewed (dismissed):
This seems fine :)

fkokosinski reviewed (dismissed):
LGTM from the perspective of RISC-V arch

andyross reviewed (commented):
Just to point out: this is not structured as a change to the interrupt architecture, it's a **new** API that some devices will use, with most continuing to use the old stuff.  The same is true of @rakons's rework to the isr table generator too, I believe.  And almost certainly @bjarki-trackunit will play the same trick.

So there's a real risk that if we wind the clock forward a few months, we'll find Zephyr has **four** different, incompatible interrupt architectures, with overlapping capability, likely some inconsistent level of interoperability, and will evolve in various stages of bitrot depending on which platforms pick them up.  That sounds worse and not better.

For the special case of the @rakons changes to the ISR table, there's an argument to merge anyway as it fixes a bug that is holding back LTO (a great feature we all want badly).  I'm not sure that this one rises to that level of impact.

Not a -1, as this seems like good code.  But I do worry that designwise we're making things worse and not better.  We need a path to get back to **one** architecture for interrupt handling that works for everyone, and every change like this makes that path longer.

ycsin commented:
Thanks for the review!

> with most continuing to use the old stuff

Would it help if all multilevel interrupt drivers are updated to use the new implementation?

rakons commented:
@andyross - I would not be so pessimistic. My implementation (#66392) is meant to be compatible with the current implementation at the source code level. When more architectures are going to be supported then the user would not be able to notice which implementation is used at the bottom. It is true - that interrupt system was not touched for eons. And now there is a lot of movement. The implementation of the controller based on #67583 would make the current (with all of the changes in #66392) obsolete, so there is nothing to worry about having 4 different implementations in the future. For the time being, from API perspective, I see only 2.

nashif commented:
> Would it help if all multilevel interrupt drivers are updated to use the new implementation?

yes, this is how we should do deprecation, calling and existing implementation "legacy" and introducing a new one leaving all users with the "legacy" approach and without a plan is something we do not want to go though. A deprecation in the tree usually triggers a CI failure, here you are deprecating via Kconfig which does not.

When we deprecate something, we want to move all users to the new feature and leave the deprecation warnings ONLY to OOT tree users.

nashif reviewed (dismissed):
need to figure out the deprecation and making current implementation legacy and ending up with 2 implementation in the tree without a plan to remove the "legacy" one.

nashif commented on include/zephyr/devicetree/interrupt_controller.h:
```diff
@@ -0,0 +1,60 @@
+ * @brief Interrupt controller devicetree macro public API header file.
+ */
+
+/*
```
This needs to be at the very top

ycsin replied:
fixed, thanks

ycsin committed:
drivers: interrupt_controller: irq_steer: use new multilevel irq impl

Update the NXP's irq_steer driver to use the new multi-level
interrupt implementation.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc_dw_ace: convert to use instance-based macros

Define the DRV_COMPAT and use instance-based device DT macros.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: update to use multi-level API

Update these multi-level interrupt drivers to use the new API.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: drivers: build_all: intc: add a common build-only test

Add a generic build-only test for:
 - intc_nxp_irqsteer
 - intc_cavs
 - intc_rv32m1_intmux
 - intc_dw_ace
 - intc_dw

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

henrikbrixandersen commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -14,6 +14,23 @@ the :ref:`release notes<zephyr_3.6>`.
 Required changes
 ****************
 
+Architecture
+============
```
This needs to go into the migration guide for v3.7.0 instead.

ycsin replied:
rebased

henrikbrixandersen commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -102,5 +102,19 @@ Userspace
 Architectures
 *************
 
+* The static auto-generation of the multilevel interrupt controller lookup table has been
+  deprecated, and will be compiled only when the new compatibility Kconfig:
+  :kconfig:option:`CONFIG_LEGACY_MULTI_LEVEL_TABLE_GENERATION` is enabled, which will eventually
+  be removed in the coming releases.
+
+  Multi-level interrupt controller drivers should be updated to use the newly created
+  ``IRQ_PARENT_ENTRY_DEFINE`` macro to register itself with the new multi-level interrupt
+  architecture. To make the macro easier to use, ``INTC_INST_ISR_TBL_OFFSET`` macro is made to
+  deduce the software ISR table offset for a given driver instance, for pseudo interrupt controller
+  child, use the ``INTC_CHILD_ISR_TBL_OFFSET`` macro instead. New devicetree macros
+  (``DT_INTC_GET_AGGREGATOR_LEVEL`` & ``DT_INST_INTC_GET_AGGREGATOR_LEVEL``) have been added
+  for an interrupt controller driver instance to pass its aggregator level into the
+  ``IRQ_PARENT_ENTRY_DEFINE`` macro.
+
```
Shouldn't this go under https://docs.zephyrproject.org/latest/releases/migration-guide-3.7.html#interrupt-controller instead of https://docs.zephyrproject.org/latest/releases/migration-guide-3.7.html#architectures ?

ycsin replied:
makes sense, fixed, thanks

ycsin committed:
tests: drivers: intc: verify registration of controllers

Test the registration of interrupt controller with the
intc_table and verify that thru the sw_isr_table APIs.

Verified registration with all 3 interrupt levels.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
force pushed again to adopt@henrikbrixandersen 's suggestion, added a new testcase and fixed an issue in the handling of 3rd level interrupt in the implementation while working on the test

ycsin commented:
Excluded `xtensa` from the test as its `CONFIG_NUM_IRQS` can't be modified, which is required by this test, opened #69579

ycsin commented:
rebased, update board names to hwmv2

dcpleung commented:
Just wondering how much have this been tested on actual hardware?

ycsin commented:
> Just wondering how much have this been tested on actual hardware?

It's been a while, to be on the safe side I'd say this is not tested on a real hardware with full blown application.

Unfortunately testing stuff inside meta hasn't been easy and straightforward, I usually rely on upstream Zephyr tools for this.

The following diff will enable both of the UARTs, one on each PLIC in the `riscv32_virtual` renode simulated board, configure the `log_backend_uart` to output on both UART ports, and run a modified `hello_world` sample that uses log:

```diff
(.venv) ycsin@LAPTOP-ROG:~/zephyrproject/zephyr$ git diff --no-prefix 
diff --git boards/renode/riscv32_virtual/riscv32_virtual.dts boards/renode/riscv32_virtual/riscv32_virtual.dts
index 326b9757518..de2814f151a 100644
--- boards/renode/riscv32_virtual/riscv32_virtual.dts
+++ boards/renode/riscv32_virtual/riscv32_virtual.dts
@@ -13,13 +13,23 @@
        compatible = "renode,riscv32-virtual";
 
        chosen {
+               zephyr,log-uart = &log_uarts;
                zephyr,console = &uart0;
                zephyr,shell-uart = &uart0;
                zephyr,flash = &flash0;
                zephyr,sram = &sram0;
        };
+
+       log_uarts: log_uarts {
+               compatible = "zephyr,log-uart";
+               uarts = <&uart0 &uart1>;
+       };
 };
 
 &uart0 {
        status = "okay";
 };
+
+&uart1 {
+       status = "okay";
+};
diff --git boards/renode/riscv32_virtual/support/riscv32_virtual.resc boards/renode/riscv32_virtual/support/riscv32_virtual.resc
index 87e327287b6..60ef8e8b8ef 100644
--- boards/renode/riscv32_virtual/support/riscv32_virtual.resc
+++ boards/renode/riscv32_virtual/support/riscv32_virtual.resc
@@ -8,6 +8,7 @@ mach create $name
 machine LoadPlatformDescription $ORIGIN/riscv32_virtual.repl
 
 showAnalyzer uart0
+showAnalyzer uart1
 cpu PerformanceInMips 4
 
 macro reset
diff --git samples/hello_world/prj.conf samples/hello_world/prj.conf
index b2a4ba59104..3c81a95875c 100644
--- samples/hello_world/prj.conf
+++ samples/hello_world/prj.conf
@@ -1 +1,2 @@
 # nothing here
+CONFIG_LOG=y
diff --git samples/hello_world/src/main.c samples/hello_world/src/main.c
index c550ab461cb..c33dd961cbf 100644
--- samples/hello_world/src/main.c
+++ samples/hello_world/src/main.c
@@ -5,10 +5,17 @@
  */
 
 #include <stdio.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+
+LOG_MODULE_REGISTER(main);
 
 int main(void)
 {
-       printf("Hello World! %s\n", CONFIG_BOARD_TARGET);
+       do {
+               LOG_INF("Hello World! %s", CONFIG_BOARD_TARGET);
+               k_msleep(1000);
+       } while (true);
 
        return 0;
 }
```

output:

```
(.venv) ycsin@LAPTOP-ROG:~/zephyrproject$ west build -b riscv32_virtual -p auto -t run zephyr/samples/hello_world/
-- west build: running target run
[0/1] cd /home/ycsin/zephyrproject/build && /usr/bin/renode --disable-xwt --port -2 --p...sin/zephyrproject/zephyr/boards/renode/riscv32_virtual/support/riscv32_virtual.resc; s'
14:16:17.2314 [INFO] Loaded monitor commands from: /opt/renode/scripts/monitor.py
14:16:17.3050 [INFO] Including script: /home/ycsin/zephyrproject/zephyr/boards/renode/riscv32_virtual/support/riscv32_virtual.resc
14:16:17.3329 [INFO] System bus created.
14:16:18.2501 [WARNING] Undefined instruction set: _.
14:16:18.2503 [WARNING] Undefined instruction set: Z.
14:16:18.2504 [WARNING] Undefined instruction set: R.
14:16:18.2505 [WARNING] Undefined instruction set: _.
14:16:18.2505 [WARNING] Undefined instruction set: Z.
14:16:18.2507 [WARNING] Undefined instruction set: E.
14:16:18.2508 [WARNING] Undefined instruction set: N.
14:16:18.2508 [WARNING] Undefined instruction set: E.
14:16:18.9268 [INFO] sysbus: Loading segment of 38404 bytes length at 0x80000000.
14:16:18.9392 [INFO] sysbus: Loading segment of 164 bytes length at 0x80009604.
14:16:18.9393 [INFO] sysbus: Loading segment of 4 bytes length at 0x800096A8.
14:16:18.9395 [INFO] sysbus: Loading segment of 6096 bytes length at 0x80400000.
14:16:18.9632 [INFO] cpu: Setting PC value to 0x80000000.
14:16:19.0319 [INFO] RISCV32-Virtual: Machine started.
14:16:19.1281 [WARNING] plic0: Unhandled write to offset 0x200000, value 0x0.
14:16:19.1552 [WARNING] plic1: Unhandled write to offset 0x200000, value 0x0.
14:16:19.2098 [INFO] uart0: [host: 0.89s (+0.89s)|virt: 9.2ms (+9.2ms)] *** Booting Zephyr OS build v3.6.0-1991-gd0f31ee48b68 ***
14:16:19.2156 [INFO] uart1: [host: 0.89s (+0.89s)|virt: 11.4ms (+11.4ms)] *** Booting Zephyr OS build v3.6.0-1991-gd0f31ee48b68 ***
14:16:19.2275 [INFO] uart0: [host: 0.91s (+17.78ms)|virt: 14.9ms (+5.7ms)] [00:00:00.000,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:19.2390 [INFO] uart1: [host: 0.92s (+23.4ms)|virt:  18.3ms (+6.9ms)] [00:00:00.000,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:25.6857 [INFO] uart0: [host:   7.36s (+6.46s)|virt:     2.01s (+2s)] [00:00:01.010,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:25.6941 [INFO] uart1: [host:  7.37s (+6.46s)|virt:      2.02s (+2s)] [00:00:01.010,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:28.9012 [INFO] uart0: [host:  10.58s (+3.22s)|virt:  3.02s (+1.01s)] [00:00:02.020,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:28.9100 [INFO] uart1: [host: 10.59s (+3.22s)|virt:   3.03s (+1.01s)] [00:00:02.020,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:32.0709 [INFO] uart0: [host:  13.75s (+3.17s)|virt:  4.03s (+1.01s)] [00:00:03.030,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:32.0793 [INFO] uart1: [host: 13.76s (+3.17s)|virt:   4.04s (+1.01s)] [00:00:03.030,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:35.2433 [INFO] uart0: [host:  16.92s (+3.17s)|virt:  5.04s (+1.01s)] [00:00:04.040,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:35.2508 [INFO] uart1: [host: 16.93s (+3.17s)|virt:   5.05s (+1.01s)] [00:00:04.040,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:38.3907 [INFO] uart0: [host:  20.07s (+3.15s)|virt:  6.05s (+1.01s)] [00:00:05.050,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
14:16:38.3990 [INFO] uart1: [host: 20.08s (+3.15s)|virt:   6.06s (+1.01s)] [00:00:05.050,000] [0m<inf> main: Hello World! riscv32_virtual/riscv_virtual_renode[0m
```

if the `logLevel` is set to `-1` (noisy), there will be traces of the PLICs getting configured during init.

ycsin commented:
ping @nashif could you please take a look agian? Thanks

andyross reviewed (approved):
I still worry that the interrupt layers are being pulled in too many directions.  But it's been a few months now and it's probably time to start merging stuff.  For sure this is an improvement on what we have.

ycsin commented:
@nashif what else should I do to move this forward?

cfriedt commented on arch/common/CMakeLists.txt:
```diff
@@ -22,6 +22,10 @@ zephyr_library_sources_ifdef(
   multilevel_irq.c
   )
 
+if(CONFIG_MULTI_LEVEL_INTERRUPTS)
+  zephyr_library_sources_ifdef(CONFIG_LEGACY_MULTI_LEVEL_TABLE_GENERATION multilevel_irq_legacy.c)
+endif()
```
Due to the `depends on`, the surrounding condition can be removed.
```suggestion zephyr_library_sources_ifdef(CONFIG_LEGACY_MULTI_LEVEL_TABLE_GENERATION multilevel_irq_legacy.c)
```

ycsin replied:
Fixed

cfriedt reviewed (dismissed):
LGTM, although you will likely want to rebase. I hope @andyross will re-approve.

andyross reviewed (dismissed):
Still no complaints, though this will cross with the new interrupt controller in #72766 and annoy me a bit when it merges.

ycsin commented:
The CI error is probably unrelated, I have created #72979

andyross reviewed (dismissed):
Refresh +1

ycsin committed:
drivers: intc: VPR CLIC should make sure GEN_IRQ_VECTOR_TABLE is enabled

VPR cores CLIC supports vectored mode only. Select
`GEN_IRQ_VECTOR_TABLE` such that it can't be disabled.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
test: gen_isr_table: exclude NRFX_CLIC from riscv_no_direct

VPR cores CLIC supports vectored mode only and can't be
disabled. Filter out `CONFIG_NRFX_CLIC` to prevent build
error.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Uh.. GitHub automatically removed some reviewers as there's not enough space, but feel free to review

ycsin commented:
Reran the doc build jobs and now it passed, ping @andyross @dcpleung @nashif
